app/benchmark_ibc_update_client_test.go

```diff
@@ -0,0 +1,516 @@
+	testCases := []struct {
+		count, numberOfValidators int
+	}{
+		{count: 6_000, numberOfValidators: 2},
+		{count: 3_000, numberOfValidators: 10},
+		{count: 2_000, numberOfValidators: 25},
+		{count: 1_000, numberOfValidators: 50},
+		{count: 500, numberOfValidators: 75},
+		{count: 500, numberOfValidators: 100},
+		{count: 500, numberOfValidators: 125},
+		{count: 500, numberOfValidators: 150},
+		{count: 500, numberOfValidators: 175},
+		{count: 500, numberOfValidators: 200},
+		{count: 500, numberOfValidators: 225},
+		{count: 500, numberOfValidators: 250},
+		{count: 500, numberOfValidators: 300},
+		{count: 500, numberOfValidators: 400},
+		{count: 500, numberOfValidators: 500},
+	}
+	for _, testCase := range testCases {
+		b.Run(fmt.Sprintf("number of validators: %d", testCase.numberOfValidators), func(b *testing.B) {
+			benchmarkIBC_ProcessProposal_Update_Client(b, testCase.numberOfValidators, testCase.count)
+		})
+	}
+}
+
+func benchmarkIBC_ProcessProposal_Update_Client(b *testing.B, numberOfValidators, count int) {
+	testApp, rawTxs := generateIBCUpdateClientTransaction(b, numberOfValidators, count)
+
+	blockData := &tmproto.Data{
+		Txs: rawTxs,
+	}
+	prepareProposalRequest := types.RequestPrepareProposal{
+		BlockData: blockData,
+		ChainId:   testApp.GetChainID(),
+		Height:    10,
+	}
+
+	prepareProposalResponse := testApp.PrepareProposal(prepareProposalRequest)
+
+	processProposalRequest := types.RequestProcessProposal{
+		BlockData: prepareProposalResponse.BlockData,
+		Header: tmproto.Header{
+			Height:   10,
+			DataHash: prepareProposalResponse.BlockData.Hash,
+			ChainID:  testutil.ChainID,
+			Version: tmprotoversion.Consensus{
+				App: testApp.AppVersion(),
+			},
+		},
+	}
+
+	b.ResetTimer()
+	resp := testApp.ProcessProposal(processProposalRequest)
+	b.StopTimer()
+	require.Equal(b, types.ResponseProcessProposal_ACCEPT, resp.Result)
+
+	b.ReportMetric(float64(b.Elapsed().Nanoseconds()), "process_proposal_time(ns)")
+	b.ReportMetric(float64(len(prepareProposalResponse.BlockData.Txs)), "number_of_transactions")
+	b.ReportMetric(float64(len(rawTxs[0])), "transactions_size(byte)")
+	b.ReportMetric(calculateBlockSizeInMb(prepareProposalResponse.BlockData.Txs), "block_size(mb)")
+	b.ReportMetric(float64(calculateTotalGasUsed(testApp, rawTxs)), "total_gas_used")
+	b.ReportMetric(float64(numberOfValidators), "number_of_validators")
+	b.ReportMetric(float64(2*numberOfValidators/3), "number_of_verified_signatures")
+}
+
+// generatePayForBlobTransactions creates a test app then generates an IBC
+// update client transaction with the specified number of validators.
+// Note: the number of the verified signatures is: 2 * numberOfValidators / 3
+func generateIBCUpdateClientTransaction(b *testing.B, numberOfValidators int, numberOfMessages int) (*app.App, [][]byte) {
+	account := "test"
+	testApp, kr := testutil.SetupTestAppWithGenesisValSet(app.DefaultConsensusParams(), account)
```

Comment:
do you know if this is using an in memory store for the iavl? if so, do you think we can try quickly with one that uses a local store?

Reply:
it is using in memory db. I switched to using levelDb. For example, `prepareProposal` for 8mb blocks of PFBS:
```
15000_transactions_of_300_bytes-16         	       1	2374515000 ns/op	         6.239 block_size(mb)	     10318 number_of_transactions	2374515000 prepare_proposal_time(ns)	988490895000 total_gas_used	       703.0 transactions_size(byte)

10000_transactions_of_500_bytes-16         	       1	1600060416 ns/op	         5.035 block_size(mb)	      6331 number_of_transactions	1600060416 prepare_proposal_time(ns)	439343930000 total_gas_used	       903.0 transactions_size(byte)

6000_transactions_of_1000_bytes-16         	       9	 112259949 ns/op	         5.809 block_size(mb)	      4566 number_of_transactions	1010339542 prepare_proposal_time(ns)	158174358000 total_gas_used	      1403 transactions_size(byte)

3000_transactions_of_5000_bytes-16         	1000000000	         0.5341 ns/op	         7.188 block_size(mb)	      1413 number_of_transactions	 534147542 prepare_proposal_time(ns)	39550179000 total_gas_used	      5403 transactions_size(byte)

1000_transactions_of_10000_bytes-16        	1000000000	         0.2072 ns/op	         7.470 block_size(mb)	       758.0 number_of_transactions	 207167333 prepare_proposal_time(ns)	4397393000 total_gas_used	     10403 transactions_size(byte)

500_transactions_of_50000_bytes-16         	1000000000	         0.1129 ns/op	         7.441 block_size(mb)	       155.0 number_of_transactions	 112946667 prepare_proposal_time(ns)	1100446500 total_gas_used	     50406 transactions_size(byte)

100_transactions_of_100000_bytes-16        	1000000000	         0.04872 ns/op	         7.368 block_size(mb)	        77.00 number_of_transactions	  48722375 prepare_proposal_time(ns)	  44369300 total_gas_used	    100406 transactions_size(byte)

100_transactions_of_200000_bytes-16        	1000000000	         0.04758 ns/op	         7.260 block_size(mb)	        38.00 number_of_transactions	  47575083 prepare_proposal_time(ns)	  44369300 total_gas_used	    200406 transactions_size(byte)

50_transactions_of_300000_bytes-16         	1000000000	         0.04190 ns/op	         7.161 block_size(mb)	        25.00 number_of_transactions	  41901375 prepare_proposal_time(ns)	  11202150 total_gas_used	    300406 transactions_size(byte)

50_transactions_of_400000_bytes-16         	1000000000	         0.03528 ns/op	         7.254 block_size(mb)	        19.00 number_of_transactions	  35280292 prepare_proposal_time(ns)	  11202150 total_gas_used	    400406 transactions_size(byte)

30_transactions_of_500000_bytes-16         	1000000000	         0.03477 ns/op	         7.157 block_size(mb)	        15.00 number_of_transactions	  34773666 prepare_proposal_time(ns)	   4085490 total_gas_used	    500406 transactions_size(byte)

10_transactions_of_1000000_bytes-16        	1000000000	         0.02764 ns/op	         6.678 block_size(mb)	         7.000 number_of_transactions	  27640250 prepare_proposal_time(ns)	    483230 total_gas_used	   1000406 transactions_size(byte)

5_transactions_of_2000000_bytes-16         	1000000000	         0.03108 ns/op	         5.723 block_size(mb)	         3.000 number_of_transactions	  31078500 prepare_proposal_time(ns)	    131790 total_gas_used	   2000406 transactions_size(byte)

3_transactions_of_3000000_bytes-16         	1000000000	         0.03043 ns/op	         5.723 block_size(mb)	         2.000 number_of_transactions	  30429959 prepare_proposal_time(ns)	     52716 total_gas_used	   3000409 transactions_size(byte)

3_transactions_of_4000000_bytes-16         	1000000000	         0.03236 ns/op	         3.815 block_size(mb)	         1.000 number_of_transactions	  32363250 prepare_proposal_time(ns)	     52716 total_gas_used	   4000409 transactions_size(byte)

2_transactions_of_5000000_bytes-16         	1000000000	         0.02360 ns/op	         4.769 block_size(mb)	         1.000 number_of_transactions	  23603458 prepare_proposal_time(ns)	     26358 total_gas_used	   5000409 transactions_size(byte)

2_transactions_of_6000000_bytes-16         	1000000000	         0.04490 ns/op	         5.722 block_size(mb)	         1.000 number_of_transactions	  44903833 prepare_proposal_time(ns)	     26358 total_gas_used	   6000409 transactions_size(byte)
```

which is almost the same. Want me to run other benchmarks too using levelDb?